src/parser/kin-parser.js

```diff
@@ -307,55 +437,111 @@ class KinParser extends Parser {
+            prevVecColor = listElements1.color
           } else {
-            let [ label2, position2, color2 ] = parseListElm(line2)
+            let listElements2 = parseListElm(line2)
 
-            if (label2 === '"') {
-              label2 = prevVecLabel
+            if (listElements2.label === '"') {
+              listElements2.label = prevVecLabel
             } else {
-              prevVecLabel = label2
+              prevVecLabel = listElements2.label
             }
 
-            if (color2 === undefined) {
-              color2 = vecDefaultColor
+            if (listElements2.color === undefined) {
+              listElements2.color = vecDefaultColor
             }
 
-            vecLabel2.push(label2)
-            vecPosition2.push(...position2)
-            vecColor2.push(...color2)
+            vecLabel2.push(listElements2.label)
+            vecPosition2.push(...listElements2.position)
+            vecColor2.push(...listElements2.color)
 
-            prevVecPosition = position2
-            prevVecColor = color2
+            prevVecPosition = listElements2.position
+            prevVecColor = listElements2.color
           }
         } else if (isBallList) {
+          // {cb arg A   1   1.431 -106.80} r=1.431  39.085, 8.083, 22.182
+
+          let listElements = parseListElm(line)
 
-        // TODO
+          if (listElements.label === '"') {
+            listElements.label = prevBallLabel
+          } else {
+            prevBallLabel = listElements.label
+          }
+
+          if (listElements.radius === undefined) {
+            listElements.radius = 1  // temporary default radius
+          }
+
+          if (listElements.color === undefined) {
+            listElements.color = ballDefaultColor
+          }
+
+          ballLabel.push(listElements.label)
+          ballRadius.push(listElements.radius)
+          ballPosition.push(...listElements.position)
+          ballColor.push(...listElements.color)
+        } else if (isRibbonList) {
+          let listElements = parseListElm(line)
 
+          if (listElements.label === '"') {
+            listElements.label = prevRibbonPointLabel
+          } else {
+            prevRibbonPointLabel = listElements.label
+          }
+
+          if (listElements.color === undefined) {
+            listElements.color = ribbonListDefaultColor
+          }
+
+          if (ribbonTriPointPositions.length < 9) {
+            ribbonTriPointLabels.push(listElements.label)
+            ribbonTriPointPositions.push(...listElements.position)
+            ribbonTriPointColors.push(...listElements.color)
+          } else {
+            ribbonPointLabelArray.push(...ribbonTriPointLabels)
+            ribbonPointPositionArray.push(...ribbonTriPointPositions)
+            ribbonPointColorArray.push(...ribbonTriPointColors)
+            ribbonTriPointLabels.shift()
+            ribbonTriPointLabels.push(listElements.label)
+            ribbonTriPointPositions.shift()
```

Comment:
Array.shift() is a very expensive operation, never use it in performance critical sections